### ISSUE TITLE
Filtering exceptions in BeforeSend is not obvious for AggregateExceptions

### ISSUE THREAD

**Issue by georgiosd** [getsentry/sentry-dotnet#3751]
### Package

Sentry

### .NET Flavor

.NET

### .NET Version

6.0.0

### OS

Windows

### SDK Version

3.34.0

### Steps to Reproduce

1. Override `OnBeforeSend` to return null when event has exception of type `SomeException`.
2. Throw `SomeException` from a `Task.Run` func.

### Expected Result

Sentry filters out the exception.

### Actual Result

Sentry still sends the exception "home".

**Comment by georgiosd** [comment]
I'm not 100% sure the repro is accurate - AFAIK it's the same conceptually with the real repro (below) but I couldn't get it to reproduce the same behavior.

"Real" repro:

1. Register a background service.
2. In `ExecuteAsync` add a small delay and then throw `SomeException`

**Comment by bricefriha** [comment]
@georgiosd Thanks for reporting this, we're taking a look

**Comment by georgiosd** [comment]
This is on an ASP.NET Core 6.0 app, hence the background service - should
have said BackgroundService.

But take a look at the issue I linked above from PuppeteerSharp - I believe
it’s thrown from a finalizer thread somewhere?

Georgios

**Comment by jamescrosswell** [comment]
@georgiosd that Pupeteer repsitory is pretty big and there's a lot going on there. It makes it quite hard to reproduce this issue.

Are you able to provide a minimal reproduction of the issue that we could look at?

**Comment by georgiosd** [comment]
Yeah, I understand. I'll spend some time and see if I can reproduce this.

**Comment by georgiosd** [comment]
Happy new year everyone!

I was struggling to get a minimal repro, so I decided to run the application locally and at least reproduce the problem, with `Debug = true`.

As it turns out, it's not a bug in the SDK, but a misleading representation in the UI. The exception is wrapped in an `AggregateException` with only exception, and so the UI shows the inner exception alone.

![Image](https://github.com/user-attachments/assets/f749dce7-bd8d-440d-b0cb-2d7e321aef5a)

Whom do we need to ping to somehow label this was unwrapped?

**Comment by bruno-garcia** [comment]
Is this on SaaS? If so, could you share a link to the event?

If Self-Hosted Sentry, I wonder what version are you on. If there are 2 exception, 1 being the `AggregateException` it should show some UI with the parent-child relationship.

**Comment by georgiosd** [comment]
Here you  go: https://zerotomvp.sentry.io/issues/6201474386/?project=5236638

**Comment by georgiosd** [comment]
Sort of. 

Visually, it's great that the UI will unwrap and you can see the actual exception and not the wrapper.

Alas, there's no way to figure out that this was done so translating back to code you end up with issues like mine.

Would it be an acceptable addition to add some sort of flag that there was a wrapper and you unwrapped it, so to speak? :)

**Comment by bruno-garcia** [comment]
> Visually, it's great that the UI will unwrap and you can see the actual exception and not the wrapper.

With `KeepAggregateException` you should see clearly that there's the actual exception under the aggregate. Could you give that a try? We could add some tag but not sure that'll be a better UX IMO

**Comment by jamescrosswell** [comment]
> With KeepAggregateException you should see clearly that there's the actual exception under the aggregate. Could you give that a try? 

@bruno-garcia `KeepAggregateException` was [removed in the 4.0.0 release](https://github.com/getsentry/sentry-dotnet/blob/7024f37efa663bbf054913b5a4659b9b407ab47d/CHANGELOG.md?plain=1#L639). 

> Alas, there's no way to figure out that this was done so translating back to code you end up with issues like mine.

@georgiosd a record of this may be [stored in the mechanism](https://github.com/getsentry/sentry-dotnet/blob/d1e5efcdf3af763ad49f11cd2426cc14a315a901/src/Sentry/Internal/MainExceptionProcessor.cs#L211) but I'm not sure how/why you'd use that. 

If what you're trying to do is filter certain exceptions, can't this be done from the `OnBeforeSend` hook? Either an AggregateException is passed as an argument to that hook (in which case you could inspect the AggregateException.InnerExceptions to see if it's one you want to filter out) or a flattened exception is passed to that hook (in which case you could check the type directly). Or am I missing something?

**Comment by georgiosd** [comment]
@jamescrosswell If you read back to why I created the ticket in the first place, is because I thought that `BeforeSend` is not working correctly in this case.

Now why did I think that? Because I looked at the UI, and wrote code that corresponds to what I saw, to filter out the exception. 

Or so I thought - because the exception was wrapped in an `AggregateException` so the code didn't work.

At the end of the day it depends on how you "read" that. We can either say "I should have known", or that the UI should somehow indicate the processing Sentry did. 😊

**Comment by jamescrosswell** [comment]
> [@jamescrosswell](https://github.com/jamescrosswell) If you read back to why I created the ticket in the first place, is because I thought that `BeforeSend` is not working correctly in this case.

Ah, OK. Sorry, got lost down the rabbit hole.

So as it stands, are we able to reproduce the behaviour that you were originally describing (`OnBeforeSend` not being triggered for unhandled exceptions in a task)? So do we know/believe there is a problem in the Sentry SDK?

**Comment by georgiosd** [comment]
I wasn't trying to say you didn't read the rabbit hole, I get it - just saying the journey kinda describes the issue at hand. 😃

The SDK is working perfectly as far as I can tell, just wish the UI kept the tiny bit of information somewhere so I don't run into this again.

**Comment by jamescrosswell** [comment]
Apologies this took so long. Whilst I don't think there's anything that needs to change in the SDK, hopefully with improved documentation we can avoid such confusion in the future. 

I've added some additional context in the docs for the SetBeforeSend options:
- https://github.com/getsentry/sentry-docs/pull/14265

@georgiosd hopefully that makes sense.

**Comment by georgiosd** [comment]
@jamescrosswell I still think the UI should be enhanced to be accurate on what it caught, so then I know what code I need to write to handle it. Having to guess if it was from an aggregate exception or always checking for it is a nuisance🤷‍♂️

**Comment by jamescrosswell** [comment]
@georgiosd this is what I'm seeing in Sentry when an AggregateException occurs:

<img width="1606" height="761" alt="Image" src="https://github.com/user-attachments/assets/9be99e68-accc-4e56-ae08-b32607157a86" />

Is that not what you're seeing or do you reckon this can be improved somehow?

**Comment by georgiosd** [comment]
This is great, but no, not what I was seeing. I am guessing it's not shown this way for unobserved task exceptions?

**Comment by jamescrosswell** [comment]
I'm not to sure. If you could provide a code sample that demonstrates the behaviour you're seeing, I could look into it further.
